onnxruntime/core/providers/cuda/tensor/gather_elements_impl.cu

```diff
@@ -1,140 +1,291 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "core/providers/cuda/tensor/gather_elements_impl.h"
+#include "core/providers/cuda/tensor/scatter_elements_impl.h"
+#ifdef ENABLE_TRAINING
+#include "orttraining/training_ops/cuda/tensor/gather_elements_grad_impl.h"
```

Comment:
I am trying to understand the bug " but the algorithm had bug that will fail to handle case when dim values along axis and the next outer axis are same. The code would coalesce these dims but it should not. And there is no UT that cover such cases.", could you please explain the case in samples, or you can point me which test case cover that one?

Reply:
Check this again, the bug is not from https://github.com/microsoft/onnxruntime/pull/2628, but from https://github.com/microsoft/onnxruntime/pull/5511 (author is me...). It will coalesce shape(input)=[2,2,2], shape(indices)=[2,2,2], axis=1 to shape(input)=[4,2], shape(indices)=[4,2], axis=0, which is wrong. I think the original PR just handled very strict cases, but failed to coalesce some other cases. Then 5511 tried to support more cases but introduced the bug.